OpenMPI: Add version 5.0.0rc9 support

amd-toolchain-support opened the pull request:
OpenMPI 5 contains significant performance improvements over OpenMPI 4 particularly for shared memory communicators, this is important for nodes with large core counts.  The release date of the final version is uncertain but our testing shows good stability and performance improvements with the latest release candidate.

Also included are a version update and fix for building from git master for the pmix package (openpmix).

matooley committed:
OpenMPI: Add version 5.0.0rc9

haampie committed:
gcc: drop target bootstrap flags for aarch64 (#33813)

See https://github.com/spack/spack/issues/31184

GCC bootstrap logic adds `-mpcu` for libatomic (iirc), which conflicts
with the `-march` flag we provide.

rhc54 commented:
FWIW: OMPI v5 is based on PMIx v4.2.2 and PRRTE v3.0.0. 

PMIx v4.2.2 has been released. 
PRRTE v3.0.0 is in final 24-hr "last comment" hold prior to release.

cessenat committed:
octave: new version 7.3.0 (#33788)

white238 committed:
add utilities and examples variants to conduit (#33804)

tldahlgren commented on var/spack/repos/builtin/packages/openmpi/package.py:
```diff
@@ -40,6 +40,8 @@ class Openmpi(AutotoolsPackage, CudaPackage):
+
     # Current
     version(
         "4.1.4", sha256="92912e175fd1234368c8730c03f4996fe5942e7479bb1d10059405e7f2b3930d"
```
You may want to add `preferred=True` to version 4.1.4 since it is the last stable release.

amd-toolchain-support replied:
Done

tldahlgren commented on var/spack/repos/builtin/packages/openmpi/package.py:
```diff
@@ -40,6 +40,8 @@ class Openmpi(AutotoolsPackage, CudaPackage):
 
     version("main", branch="main", submodules=True)
 
+    version("5.0.0rc9", branch="main", tag="5.0.0rc9", submodules=True)
```
I believe you only want the tag here.  You already have a `main` branch version.

```suggestion
    version("5.0.0rc9", tag="5.0.0rc9", submodules=True)
```

amd-toolchain-support replied:
It needs a branch because otherwise it fails with a messages that branch 5.0.0rc9 does not exist.  However, I can use the `v5.0.x` branch I think.

tldahlgren reviewed (commented):
Confirmed the new sha256 and that the new tag exists.  Have suggested changes.

rhc54 commented:
I'm afraid I am not a Spack expert, but a couple of comments about the required fields:

* OMPI v5 requires a minimum of PMIx v4.2.2 - your current requirement seems to be "4.2", which might allow the build to incorrectly pickup v4.2.0 or v4.2.1, which are not supported.
* I don't see any entry for the required PRRTE v3.0.0 (or above) package - perhaps I am just missing it?

hppritcha commented:
we don't want 5.0.0rc's labeled as main.  i actually do nightly openmpi main spack based builds and this change would mess things up.

note we build main differently from release branches.  one important difference is we build the pmix/prrte internal versions and not external ones.  The reasons for this is the shas for these submodules are being moved in a dynamic fashion that doesn't necessarily track particular releases of these two products.

RemiLacroix-IDRIS committed:
pandoc: add version 2.19.2 (#33811)

glennpj committed:
Cran updates (#33819)

* add version 1.7-20 to r-ade4
* add version 1.1-13 to r-adephylo
* add version 0.3-20 to r-adespatial
* add version 1.2-0 to r-afex
* add version 0.8-19 to r-amap
* add version 0.1.8 to r-aplot
* add version 2.0.8 to r-biasedurn
* add version 2.4-4 to r-bio3d
* add version 1.30.19 to r-biocmanager
* add version 1.2-9 to r-brobdingnag
* add version 0.4.1 to r-bslib
* add version 3.7.3 to r-callr
* add version 3.1-1 to r-car
* add version 0.3-62 to r-clue
* add version 1.2.0 to r-colourpicker
* add version 1.8.1 to r-commonmark
* add version 0.4.3 to r-cpp11
* add version 1.14.4 to r-data-table
* add version 1.34 to r-desolve
* add version 2.4.5 to r-devtools
* add version 0.6.30 to r-digest
* add version 0.26 to r-dt
* add version 1.7-12 to r-e1071
* add version 1.8.2 to r-emmeans
* add version 1.1.16 to r-exomedepth
* add version 0.1-8 to r-expint
* add version 0.4.0 to r-fontawesome
* add version 1.5-2 to r-fracdiff
* add version 1.10.0 to r-future-apply
* add version 3.0.1 to r-ggmap
* add version 3.4.0 to r-ggplot2
* add version 2.1.0 to r-ggraph
* add version 0.6.4 to r-ggsignif
* add version 0.6-7 to r-gmp
* add version 2022.10-2 to r-gparotation
* add version 0.8.3 to r-graphlayouts
* add version 1.8.8 to r-grbase
* add version 2.1-0 to r-gstat
* add version 0.18.6 to r-insight
* add version 1.3.1 to r-irkernel
* add version 1.8.3 to r-jsonlite
* add version 1.7.0 to r-lava
* add version 1.1-31 to r-lme4
* add version 5.6.17 to r-lpsolve
* add version 5.5.2.0-17.9 to r-lpsolveapi
* add version 1.2.9 to r-mapproj
* add version 3.4.1 to r-maps
* add version 1.1-5 to r-maptools
* add version 1.3 to r-markdown
* add version 6.0.0 to r-mclust
* add version 4.2-2 to r-memuse
* add version 1.8-41 to r-mgcv
* add version 1.2.5 to r-minqa
* add version 0.3.7 to r-nanotime
* add version 2.4.1.1 to r-nfactors
* add version 3.1-160 to r-nlme
* add version 2.7-1 to r-nmof
* add version 0.60-16 to r-np
* add version 2.0.4 to r-openssl
* add version 4.2.5.1 to r-openxlsx
* add version 0.3-8 to r-pbdzmq
* add version 2.0-3 to r-pcapp
* add version 0.7.0 to r-philentropy
* add version 2.0.3 to r-pkgcache
* add version 1.3.1 to r-pkgload
* add version 1.10-4 to r-polyclip
* add version 3.8.0 to r-processx
* add version 1.7.2 to r-ps
* add version 2.12.1 to r-r-utils
* add version 1.2.4 to r-ragg
* add version 3.7 to r-rainbow
* add version 0.0.20 to r-rcppannoy
* add version 0.3.3.9.3 to r-rcppeigen
* add version 0.3.12 to r-rcppgsl
* add version 1.0.2 to r-recipes
* add version 1.1.10 to r-rmutil
* add version 0.10.24 to r-rmysql
* add version 2.4.8 to r-rnexml
* add version 4.1.19 to r-rpart
* add version 1.7-2 to r-rrcov
* add version 0.8.28 to r-rsconnect
* add version 0.25 to r-servr
* add version 1.7.3 to r-shiny
* add version 3.0-0 to r-spatstat-data
* add version 3.0-3 to r-spatstat-geom
* add version 3.0-1 to r-spatstat-random
* add version 3.0-0 to r-spatstat-sparse
* add version 3.0-1 to r-spatstat-utils
* add version 1.8.0 to r-styler
* add version 3.4.1 to r-sys
* add version 1.5-2 to r-tclust
* add version 0.0.9 to r-tfmpvalue
* add version 1.2.0 to r-tidyselect
* add version 0.10-52 to r-tseries
* add version 4.2.2 to r-v8
* add version 0.5.0 to r-vctrs
* add version 2.6-4 to r-vegan
* add version 0.7.0 to r-wk
* add version 0.34 to r-xfun
* add version 1.0.6 to r-xlconnect
* add version 3.99-0.12 to r-xml
* add version 0.12.2 to r-xts
* add version 1.0-33 to r-yaimpute
* add version 2.3.6 to r-yaml
* add version 2.2.2 to r-zip
* add version 2.2-7 to r-deoptim
* add version 4.3.1 to r-ergm
* add version 0.18 to r-evaluate
* add version 1.29.0 to r-future
* add version 0.0.8 to r-ggfun
* add version 0.9.2 to r-ggrepel
* add version 1.9.0 to r-lubridate
* add version 4.10.1 to r-plotly
* add version 0.2.12 to r-rcppcctz
* add version 1.2 to r-rook
* add version 1.6-1 to r-segmented
* add version 4.2.1 to r-seurat
* add version 4.1.3 to r-seuratobject
* add version 1.0-9 to r-sf
* add version 1.5-1 to r-sp
* add version 1.8.1 to r-styler
* new package: r-timechange
* new package: r-stars
* new package: r-sftime
* new package: r-spatstat-explore

Co-authored-by: glennpj <glennpj@users.noreply.github.com>

balay committed:
xsdk: deprecate 0.6.0, and remove older deprecated releases (#33815)

xsdk-examples: deprecate @0.2.0, remove @0.1.0 in sync with older xsdk releases

MatthewLieber committed:
adding sha for 7.0 release (#33823)

Co-authored-by: Matthew Lieber <lieber.31@osu.edu>

alalazo committed:
Delete directory with vestigial configuration file (#33825)

haampie committed:
Use bfs order in spack find --deps tree (#33782)

* Use bfs order in spack find --deps tree
* fix printing tests

becker33 committed:
remove deprecated `concretization` environment key (#33774)

lpottier committed:
redis-plus-plus: added initial support (#33803)

* redis-plus-plus: added initial support
* redis-plus-plus: use cmake arg provided by Spack
* redis-plus-plus: oups tiny typo

Signed-off-by: Loïc Pottier <pottier1@llnl.gov>
Signed-off-by: Loïc Pottier <lpottier@arnor>
Co-authored-by: Loïc Pottier <lpottier@arnor>

sethrj committed:
go-bootstrap: improve error message for Monterey (#31800)

Co-authored-by: Seth R. Johnson <johnsonsr@ornl.gov>

lrtfm committed:
global: add --with-ncurses with prefix to configure (#33136)

pranith committed:
emacs: Add version 28.2 (#33255)

bvanessen committed:
Fixed a bug where the external HIP library is found in a nested directory, even on newer releases of ROCm. (#33772)

becker33 committed:
remove deprecated top-level module config (#33828)

* remove deprecated top-level module config per deprecation in 0.18

haampie committed:
remove activate/deactivate support in favor of environments (#29317)

Environments and environment views have taken over the role of `spack activate/deactivate`, and we should deprecate these commands for several reasons:

- Global activation is a really poor idea:
   - Install prefixes should be immutable; since they can have multiple, unrelated dependents; see below
   - Added complexity elsewhere: verification of installations, tarballs for build caches, creation of environment views of packages with unrelated extensions "globally activated"... by removing the feature, it gets easier for people to contribute, and we'd end up with fewer bugs due to edge cases.
- Environment accomplish the same thing for non-global "activation" i.e. `spack view`, but better.

Also we write in the docs:

```
However, Spack global activations have two potential drawbacks:

#. Activated packages that involve compiled C extensions may still
   need their dependencies to be loaded manually.  For example,
   ``spack load openblas`` might be required to make ``py-numpy``
   work.

#. Global activations "break" a core feature of Spack, which is that
   multiple versions of a package can co-exist side-by-side.  For example,
   suppose you wish to run a Python package in two different
   environments but the same basic Python --- one with
   ``py-numpy@1.7`` and one with ``py-numpy@1.8``.  Spack extensions
   will not support this potential debugging use case.
```

Now that environments are established and views can take over the role of activation
non-destructively, we can remove global activation/deactivation.

matooley committed:
OpenMPI@5.0.0rc9 fixes for PR

amd-toolchain-support committed:
Merge branch 'develop' into openmpi5rc9

rhc54 commented:
> @rhc54 There doesn't appear to be any existing dependency on prrte in this package. Is this a new dependency or is it getting shipped in release tarballs?

Yes, it is a new dependency in OMPI v5. The built-in RTE was replaced to use PRRTE starting with the v5.0.0 release. You are likely missing it because OMPI includes an embedded copy of it (just like it does with PMIx) in its tarball. @hppritcha knows the details and perhaps can provide the necessary missing Spack reference.

amd-toolchain-support commented:
Okay.  So would a sensible solution be to use the 5.0.0rc9 tarballs I have just found out are a thing?  That way we get the current prrte bundled without needing to worry about adding more logic to the recipe here.

rhc54 commented:
I think that's totally up to you, largely depending upon your objectives. If you just want an all-included result, then no need to expose openpmix or prrte dependencies - everything is already in the tarball. You can also drop the libevent and hwloc dependencies as well, if you like, as they are also included.

If you want the option to independently install/maintain the individual component packages, then exposing the dependencies is the way to go. Just note that if you expose openpmix and/or prrte as a dependency, then you must also expose libevent and hwloc to ensure that OMPI uses all external components.

Rule is: either all internal, or all external - no mixing allowed. So probably the best answer is to either add the prrte dependency, or remove them all (libevent, hwloc, pmix, and prrte) and use the embedded versions.

hppritcha reviewed (approved):
I think this is good enough for now.  if there's another rc for 5.0.0, which is not out of the realm of possibility,  the spack recipe can be easily modified.  Once 5.0.0 is released, we may need to revisit if we think there's a need to build prrte externally.

tldahlgren commented:
FWIW. I was able to build `4.1.4` but not `5.0.0rc9` on an LLNL computer.  The output from the latter was:

```
...
==> Installing openmpi-5.0.0rc9-ysbtjadbftjccxhmyovti7velb255gni
==> No binary for openmpi-5.0.0rc9-ysbtjadbftjccxhmyovti7velb255gni found: installing from source
==> No patches needed for openmpi
==> openmpi: Executing phase: 'autoreconf'
==> openmpi: Executing phase: 'configure'
==> Error: ProcessError: Command exited with status 1:
    '/var/tmp/dahlgren/spack-stage/spack-stage-openmpi-5.0.0rc9-ysbtjadbftjccxhmyovti7velb255gni/spack-src/configure' '--prefix=/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/openmpi-5.0.0rc9-ysbtjadbftjccxhmyovti7velb255gni' '--enable-shared' '--disable-silent-rules' '--disable-builtin-atomics' '--enable-static' '--enable-mpi1-compatibility' '--without-ucx' '--without-cma' '--without-verbs' '--without-psm' '--without-ofi' '--without-mxm' '--without-psm2' '--without-fca' '--without-hcoll' '--without-xpmem' '--without-knem' '--without-cray-xpmem' '--without-slurm' '--without-alps' '--without-tm' '--without-sge' '--without-lsf' '--without-loadleveler' '--disable-memchecker' '--with-libevent=/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/libevent-2.1.12-tvq52uq2xf5of4cyafznwczvajgz2ybq' '--with-pmix=/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/pmix-4.2.2-4vrtpcdkgikdhq6hifadzdrdkwqaafye' '--with-zlib=/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/zlib-1.2.11-lddqpqw7puyfni3vxtytu3og6tdfrx5j' '--with-hwloc=/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/hwloc-2.7.1-bw3gsbdwzhnxgwtdhgxlggzshabt3fak' '--disable-java' '--disable-mpi-java' '--with-gpfs=no' '--without-cuda' '--enable-wrapper-rpath' '--disable-wrapper-runpath' '--with-wrapper-ldflags=-Wl,-rpath,/usr/tce/packages/gcc/gcc-8.3.1/rh/usr/lib/gcc/x86_64-redhat-linux/8'

29 errors found in build log:
     575     libtoolize: copying file './config/ltsugar.m4'
     576     libtoolize: copying file './config/ltversion.m4'
     577     libtoolize: copying file './config/lt~obsolete.m4'
     578     autoreconf: running: /usr/bin/autoconf --include=config --force --
             warnings=all,no-obsolete,no-override
     579     autoreconf: running: /usr/bin/autoheader --include=config --force 
             --warnings=all,no-obsolete,no-override
     580     autoreconf: running: automake --add-missing --copy --force-missing
              --warnings=all,no-obsolete,no-override
  >> 581     configure.ac:211: installing './config/ar-lib'
  >> 582     configure.ac:260: installing './config/compile'
  >> 583     configure.ac:98: installing './config/config.guess'
  >> 584     configure.ac:98: installing './config/config.sub'
  >> 585     configure.ac:101: installing './config/install-sh'
  >> 586     configure.ac:101: installing './config/missing'
     587     examples/Makefile.am: installing './config/depcomp'
     588     configure.ac: installing './config/ylwrap'
     589     parallel-tests: installing './config/test-driver'
     590     autoreconf: Leaving directory `.'
     591     === Patching PGI compiler version numbers in ltmain.sh
     592     === Patching "-pthread" option for NAG compiler in ltmain.sh

     ...

     774     libtoolize: copying file 'config/ltsugar.m4'
     775     libtoolize: copying file 'config/ltversion.m4'
     776     libtoolize: copying file 'config/lt~obsolete.m4'
     777     autoreconf: running: /usr/bin/autoconf --include=config --force --
             warnings=all,no-obsolete,no-override
     778     autoreconf: running: /usr/bin/autoheader --include=config --force 
             --warnings=all,no-obsolete,no-override
     779     autoreconf: running: automake --add-missing --copy --force-missing
              --warnings=all,no-obsolete,no-override
  >> 780     configure.ac:319: installing 'config/compile'
  >> 781     configure.ac:105: installing 'config/config.guess'
  >> 782     configure.ac:105: installing 'config/config.sub'
  >> 783     configure.ac:132: installing 'config/install-sh'
  >> 784     configure.ac:132: installing 'config/missing'
     785     src/Makefile.am: installing 'config/depcomp'
     786     configure.ac: installing 'config/ylwrap'
     787     autoreconf: Leaving directory `.'
     788     === Patching PGI compiler version numbers in ltmain.sh
     789     === Patching "-pthread" option for NAG compiler in ltmain.sh
     790     === Replacing ./config/config.sub with newer version

     ...

     828     libtoolize: copying file 'confdb/ltsugar.m4'
     829     libtoolize: copying file 'confdb/ltversion.m4'
     830     libtoolize: copying file 'confdb/lt~obsolete.m4'
     831     autoreconf: running: /usr/bin/autoconf --include=confdb --force
     832     autoreconf: running: /usr/bin/autoheader --include=confdb --force
     833     autoreconf: running: automake --add-missing --copy --force-missing
  >> 834     configure.ac:50: installing 'confdb/ar-lib'
  >> 835     configure.ac:44: installing 'confdb/compile'
  >> 836     configure.ac:52: installing 'confdb/config.guess'
  >> 837     configure.ac:52: installing 'confdb/config.sub'
  >> 838     configure.ac:38: installing 'confdb/install-sh'
  >> 839     configure.ac:38: installing 'confdb/missing'
     840     Makefile.am: installing 'confdb/depcomp'
     841     autoreconf: Leaving directory `.'
     842     === running autoreconf in 'mpl' ===
     843     autoreconf: Entering directory `.'
     844     autoreconf: configure.ac: not using Gettext
     845     autoreconf: running: aclocal --force -I confdb

     ...

     853     libtoolize: copying file 'confdb/ltsugar.m4'
     854     libtoolize: copying file 'confdb/ltversion.m4'
     855     libtoolize: copying file 'confdb/lt~obsolete.m4'
     856     autoreconf: running: /usr/bin/autoconf --force
     857     autoreconf: running: /usr/bin/autoheader --force
     858     autoreconf: running: automake --add-missing --copy --force-missing
  >> 859     configure.ac:24: installing 'confdb/ar-lib'
  >> 860     configure.ac:20: installing 'confdb/compile'
  >> 861     configure.ac:30: installing 'confdb/config.guess'
  >> 862     configure.ac:30: installing 'confdb/config.sub'
  >> 863     configure.ac:15: installing 'confdb/install-sh'
  >> 864     configure.ac:15: installing 'confdb/missing'
     865     Makefile.am: installing 'confdb/depcomp'
     866     parallel-tests: installing 'confdb/test-driver'
     867     autoreconf: Leaving directory `.'
     868         Patching configure for Libtool PGI 10 fortran compiler name
     869         Patching configure for Libtool PGI version number regexps
     870         Patching configure for flang Fortran ()

     ...

     900     libtoolize: copying file 'config/ltsugar.m4'
     901     libtoolize: copying file 'config/ltversion.m4'
     902     libtoolize: copying file 'config/lt~obsolete.m4'
     903     autoreconf: running: /usr/bin/autoconf --include=config --force --
             warnings=all,no-obsolete,no-override
     904     autoreconf: running: /usr/bin/autoheader --include=config --force 
             --warnings=all,no-obsolete,no-override
     905     autoreconf: running: automake --add-missing --copy --force-missing
              --warnings=all,no-obsolete,no-override
  >> 906     configure.ac:368: installing 'config/compile'
  >> 907     configure.ac:96: installing 'config/config.guess'
  >> 908     configure.ac:96: installing 'config/config.sub'
  >> 909     configure.ac:108: installing 'config/install-sh'
  >> 910     configure.ac:108: installing 'config/missing'
     911     3rd-party/treematch/Makefile.am: installing 'config/depcomp'
     912     parallel-tests: installing 'config/test-driver'
     913     configure.ac: installing 'config/ylwrap'
     914     autoreconf: Leaving directory `.'
     915     === Patching PGI compiler version numbers in ltmain.sh
     916     === Patching "-pthread" option for NAG compiler in ltmain.sh

     ...

     1858    checking for hwloc_topology_dup... yes
     1859    
     1860    *** Configuring PMIx
     1861    checking for pmix pkg-config name... /usr/WS1/dahlgren/spack/opt/s
             pack/linux-rhel7-broadwell/gcc-8.3.1/pmix-4.2.2-4vrtpcdkgikdhq6hif
             adzdrdkwqaafye/lib/pkgconfig/pmix.pc
     1862    checking if pmix pkg-config module exists... yes
     1863    checking for pmix pkg-config cflags... error
  >> 1864    configure: error: An error occurred retrieving pmix cppflags from 
             pkg-config
```

hppritcha commented:
what happened with this PR?  why are there all these additional commits that don't have to do with Open MPI?

tldahlgren commented:
> @tldahlgren Trying to rebase this patch on the tip of develop to appease github's complaints hasn't worked like I was expecting it to. Do I need to rework this PR and force push?

All the extra commits appears to be an issue with the rebase.

alalazo committed:
Merge branch 'develop' into openmpi5rc9
